ariadne/src/Ariadne/Wallet/Backend/KeyStorage.hs

```diff
@@ -126,121 +141,165 @@ instance Exception NotARenamableItem where
  displayException NotARenamableItem =
     "This item cannot be renamed."
 
+data WalletIndexOutOfRange = WalletIndexOutOfRange Word
+  deriving (Eq, Show)
+
+instance Exception WalletIndexOutOfRange where
+  displayException (WalletIndexOutOfRange i) =
+   "The index " ++ show i ++ " is out of range."
+
 -- | Utility function that runs CatchT monad inside a StateT and rollbacks the state on failure
 runCatchInState :: Functor m => StateT s (CatchT m) a -> StateT s m (Either SomeException a)
 runCatchInState m = StateT $ \s ->
   runCatchT (runStateT m s) <&> \case
     Left e -> (Left e, s)
     Right (a, s') -> (Right a, s')
 
--- | Get the wallet index by name or using current selection.
+-- | Get the wallet HdRootId by name or using current selection.
 resolveWalletRef
   :: IORef (Maybe WalletSelection)
-  -> (CardanoMode ~> IO)
   -> WalletReference
-  -> IO Word
-resolveWalletRef walletSelRef runCardanoMode = \case
+  -> DB
+  -> IO HdRootId
```

Comment:
Maybe let's return `HdRoot` instead of `HdRootId`? So that the caller doesn't need to handle cases where `HdRootId` corresponding to `HdRoot` doesn't exist (which should be impossible).

Reply:
AFAIU, a caller can not omit handling invalid `HdRootId` case, because query functions always validate it and return `Either e Something`. It does not matter, how we get `HdRootId`: `hdRoot ^. hdRootId` or get it directly from `resolveWalletRef`.  Now I use unsafe local functions because it is known that `resolveWalletRef` always returns valid `HdRootId`.